### PR TITLE
feat: move more logic into workers

### PR DESCRIPTION
**Pull request by Julusian**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature

* **What is the current behavior?** (You can also link to an open issue here)

Meteor is sending messages from the ExternalMessageQueue in the background.
There is no need for meteor to do this work, and as it is written it will not scale very well with meteor

* **What is the new behavior (if this is a feature change)?**

The ExternalMessageQueue sending loop/timer/watcher has been moved to the 'events' thread. This thread was previously tasked with low-priority 'events' (yellow line to enps, metadata sending), so was appropriate to add some more timing insensitive logic to.
No new jobs were added, the queue remains running on its own timers and database watching. This shouldnt impact performance negatively, as nothing else in the thread is time sensitive.
As this queue is not being run as a Job, it is the one place in the worker where setInterval/setTimeout are allowed to be used.

In the future, I could see us moving more background tasks to this thread, such as the cronjob to check if there are any playlists with no rundowns, or generating/updating status messages.

Additionally, I have moved the playlist snapshot generation into the worker, so that it is an 'atomic' operation that is done inside of the playlist lock. This will ensure that the snapshot isnt taken mid-way though a save. I am unsure if this is worth the complexity it adds, so am open to discussing whether this should be kept.

* **Other information**:

**Status**

- [ ] Code documentation for the relevant parts in the code have been added/updated by the PR author
- [ ] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

**Comment by Julusian** [comment]
I need to move some tests still, and I havent tested it yet (I am a little unsure how best to do that), but the general architecture and approach can be reviewed

File: packages/job-worker/src/interval/playlist.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// import { RundownPlaylistId } from '@sofie-automation/corelib/dist/dataModel/Ids'
```
Comment (Julusian) [inline packages/job-worker/src/interval/playlist.ts:1]: I didn't mean to stage this file, it was from another experiment

**Comment by Julusian** [comment]
I have given slack based external messages a test, and they get sent properly. I do not expect to test rabbitmq, as the changes to the specific sections are not invasive enough to feel it is justified.

All that is left of this is to migrate over the unit tests

File: packages/job-worker/src/db/collections.ts [hunk 2]
```diff
@@ -92,44 +113,76 @@ export interface IDirectCollections {
 	ExternalMessageQueue: ICollection<ExternalMessageQueueObj>
 }
 
-export function getMongoCollections(client: MongoClient, dbName: string): Readonly<IDirectCollections> {
+export function getMongoCollections(
+	client: MongoClient,
+	dbName: string,
+	allowWatchers: boolean
```
Comment (nytamin) [inline packages/job-worker/src/db/collections.ts:119]: This parameter could use some documentation. What is it's purpose?
(same thing with `#allowWatchers` in `WrappedCollection`)

File: packages/job-worker/src/workers/events/child.ts [hunk 3]
```diff
@@ -81,6 +96,8 @@ export class EventsWorkerChild {
 		}
 
 		try {
+			if (this.#externalMessageQueue) this.#externalMessageQueue.invalidateCaches(data)
```
Comment (nytamin) [inline packages/job-worker/src/workers/events/child.ts:99]: -> `this.#externalMessageQueue?.invalidateCaches(data)`

File: packages/job-worker/src/db/collections.ts [hunk 4]
```diff
@@ -58,6 +60,25 @@ export interface ICollection<TDoc extends { _id: ProtectedString<any> }> {
 	replace(doc: TDoc | ReadonlyDeep<TDoc>): Promise<boolean>
 
 	bulkWrite(ops: Array<AnyBulkWriteOperation<TDoc>>): Promise<unknown>
+
+	/**
+	 * Watch the collection for changes
+	 * Note: Not available in all workers/situations
```
Comment (nytamin) [inline packages/job-worker/src/db/collections.ts:66]: This is a quite scary comment :) could you elaborate?

**Review by nytamin — APPROVED** [review]
I left a few minor comments, other than that this looks good!

File: packages/job-worker/src/db/collections.ts (hunk 2 above)
Comment (Julusian) [reply packages/job-worker/src/db/collections.ts:126]: better?

File: packages/job-worker/src/db/collections.ts (hunk 4 above)
Comment (Julusian) [reply packages/job-worker/src/db/collections.ts]: better?

**Pull request merged**